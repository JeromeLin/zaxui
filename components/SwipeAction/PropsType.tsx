interface BaseProps {
  left?: Array<Object>;
  right?: Array<Object>;
  moveDistanceRatio: number;
  moveTimeSpan: number;
  animationDuration?: number;
  offset: number;
  autoClose?: boolean;
  disabled?: boolean;
  onOpen?: () => void;
  onClose?: () => void;
}

export interface SwipeActionProps extends BaseProps {
  prefixCls?: string;
  className?: string;
}
