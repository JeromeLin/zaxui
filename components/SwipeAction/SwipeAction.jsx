import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Events from '../utils/events';
import Drag from '../Drag';

class SwipeAction extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      offsetLeft: 0,
    };
    this.isOpen = false;
    this.touchEnd = true;
    this.onDragStart = this.onDragStart.bind(this);
    this.onDragMove = this.onDragMove.bind(this);
    this.onDragEnd = this.onDragEnd.bind(this);
  }

  componentDidMount() {
    Events.on(document.body, 'touchstart', e => this.onCloseSwipe(e));
  }

  componentWillUnmount() {
    Events.off(document.body, 'touchstart', e => this.onCloseSwipe(e));
  }

  onDragStart() {
    if (this.isOpen) {
      this.touchEnd = false;
      this.close();
      return;
    }
    this.touchEnd = true;
  }

  onDragMove(event, { offsetX, offsetY }) {
    if (!this.touchEnd) return;

    const { disabled } = this.props;
    if (disabled) return;

    // 拖动距离达到上限
    const { offset } = this.props;
    const { offsetLeft } = this.state;
    const btnsLeftWidth = this.left && this.left.offsetWidth;
    const btnsRightWidth = this.right && this.right.offsetWidth;
    if (offsetX > 0 && (!btnsLeftWidth || offsetLeft >= btnsLeftWidth + offset)) return false;
    if (offsetX < 0 && (!btnsRightWidth || offsetLeft <= -btnsRightWidth - offset)) return false;

    // 判断滚屏情况
    const distanceX = Math.abs(offsetX);
    const distanceY = Math.abs(offsetY);
    if (distanceX < 5 || (distanceX >= 5 && distanceY >= 0.3 * distanceX)) return false;

    event.preventDefault();

    this.doTransition({ offsetLeft: offsetX, duration: 0 });
    return true;
  }

  onDragEnd(event, { offsetX, startTime }) {
    const { duration, moveDistanceRatio, moveTimeSpan } = this.props;
    const timeSpan = new Date().getTime() - startTime.getTime();
    const btnsLeftWidth = this.left && this.left.offsetWidth;
    const btnsRightWidth = this.right && this.right.offsetWidth;

    let distanceX = 0;
    let isOpen = false;

    if ((offsetX / btnsLeftWidth > moveDistanceRatio) || (offsetX > 0 && timeSpan <= moveTimeSpan)) {
      distanceX = btnsLeftWidth;
      isOpen = true;
    } else if ((offsetX / btnsRightWidth < -moveDistanceRatio) || (offsetX < 0 && timeSpan <= moveTimeSpan)) {
      distanceX = -btnsRightWidth;
      isOpen = true;
    }

    if (isOpen && !this.isOpen) {
      // 打开
      this.open(distanceX);
    } else if (!isOpen && this.isOpen) {
      // 关闭
      this.close();
    } else {
      // 还原
      this.doTransition({ offsetLeft: distanceX, duration });
    }
  }

  onBtnClick(e, btn) {
    e.preventDefault();
    const onClick = btn.onClick;
    if (onClick) {
      onClick(e);
    }

    if (this.props.autoClose) {
      this.close();
    }
  }

  onCloseSwipe(e) {
    if (!this.wrap) return;

    if (this.isOpen) {
      const pNode = ((node) => {
        while (node.parentNode && node.parentNode !== document.body) {
          if (node === this.wrap) {
            return node;
          }
          node = node.parentNode;
        }
      })(e.target);

      if (!pNode) {
        e.preventDefault();
        this.touchEnd = true;
        this.close();
      }
    }
  }

  open(offsetLeft) {
    const { duration, onOpen } = this.props;
    onOpen();
    this.isOpen = true;
    this.doTransition({ offsetLeft, duration });
  }

  close() {
    const { duration, onClose } = this.props;
    onClose();
    this.isOpen = false;
    this.doTransition({ offsetLeft: 0, duration });
  }

  doTransition({ offsetLeft, duration }) {
    this.setState({ offsetLeft, duration });
  }

  renderButtons(buttons, ref) {
    const prefixCls = this.props.prefixCls;

    return (buttons && buttons.length) ? (
      <div
        className={`${prefixCls}-actions-${ref}`}
        ref={(el) => { this[ref] = el; }}>
        {
          buttons.map((btn, i) => {
            const { theme, className, text } = btn;

            const classes = classnames({
              [`${prefixCls}-button`]: true,
              [`theme-${theme}`]: true,
              [className]: !!className,
            });

            return (
              <div
                key={+i}
                className={classes}
                onClick={e => this.onBtnClick(e, btn)}>
                <div className={`${prefixCls}-text`}>{text || `${ref}${i}`}</div>
              </div>
            );
          })
        }
      </div>
    ) : null;
  }

  render() {
    const { prefixCls, className, left, right, children } = this.props;
    const { offsetLeft, duration } = this.state;

    const classes = classnames({
      [`${prefixCls}`]: true,
      [className]: !!className,
    });

    const style = {
      WebkitTransitionDuration: `${duration}ms`,
      transitionDuration: `${duration}ms`,
      WebkitTransform: `translate3d(${offsetLeft}px, 0, 0)`,
      transform: `translate3d(${offsetLeft}px, 0, 0)`,
    };

    return (left.length || right.length)
      ? (
        <div className={classes} ref={(wrap) => { this.wrap = wrap; }}>
          {this.renderButtons(left, 'left')}
          {this.renderButtons(right, 'right')}
          <Drag
            onDragStart={this.onDragStart}
            onDragMove={this.onDragMove}
            onDragEnd={this.onDragEnd}>
            <div className={`${prefixCls}-content`} style={style}>
              {children}
            </div>
          </Drag>
        </div>
        )
      : children;
  }
}

SwipeAction.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  left: PropTypes.arrayOf(PropTypes.object),
  right: PropTypes.arrayOf(PropTypes.object),
  moveDistanceRatio: PropTypes.number,
  moveTimeSpan: PropTypes.number,
  duration: PropTypes.number,
  offset: PropTypes.number,
  onOpen: PropTypes.func,
  onClose: PropTypes.func,
};

SwipeAction.defaultProps = {
  prefixCls: 'za-swipeaction',
  className: null,
  left: [],
  right: [],
  moveDistanceRatio: 0.5,
  moveTimeSpan: 300,
  duration: 300,
  offset: 10,
  onOpen() {},
  onClose() {},
};

export default SwipeAction;
