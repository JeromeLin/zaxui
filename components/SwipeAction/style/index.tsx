import '../../style/';
import '../../Button/style/';
import './index.scss';
