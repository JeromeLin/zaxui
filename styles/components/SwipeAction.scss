.ui-swipeAction-wrap{
  position: relative;
  .ui-swipeAction-btn-wrap{
    position: absolute;
    height: 100%;
    right: 0;
    top: 0;
    z-index: -1;

  }
  .full-mask{
    position: fixed;
    left: 0;
    top: 0;
    bottom: 0;
    right: 0;
    z-index: 100;
    background: #fff;
    opacity: 0.2;
  }
}
.demo-btn-wrap{
  height: 100%;
}
.demo-del{
  display: inline-block;
  width: r(80);
  text-align: center;
  &.del1{
    background: #1AAD19;
    height: 100%;
  }
  &.del2{
    background: #E64340;
    height: 100%;
  }
}