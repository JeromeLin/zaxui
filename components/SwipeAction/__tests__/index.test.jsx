import React from 'react';
import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import SwipeAction from '../index';

describe('SwipeAction', () => {
  it('renders correctly', () => {
    const wrapper = render(
      <SwipeAction
        left={[
          {
            theme: 'info',
            text: '左按钮1',
            onClick: () => console.log('左按钮1'),
          },
          {
            theme: 'warning',
            text: '左按钮2',
            onClick: () => console.log('左按钮2'),
          },
        ]}
        right={[
          {
            theme: 'error',
            text: '右按钮1',
            onClick: () => console.log('右按钮1'),
          },
          {
            theme: 'success',
            text: '右按钮2',
            onClick: () => console.log('右按钮2'),
          },
        ]}>
        <div>左右都能滑动</div>
      </SwipeAction>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
