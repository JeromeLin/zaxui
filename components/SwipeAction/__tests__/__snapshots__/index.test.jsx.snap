// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwipeAction renders correctly 1`] = `
<div
  class="za-swipeaction"
>
  <div
    class="za-swipeaction-actions-left"
  >
    <div
      class="za-swipeaction-button theme-info"
    >
      <div
        class="za-swipeaction-text"
      >
        左按钮1
      </div>
    </div>
    <div
      class="za-swipeaction-button theme-warning"
    >
      <div
        class="za-swipeaction-text"
      >
        左按钮2
      </div>
    </div>
  </div>
  <div
    class="za-swipeaction-actions-right"
  >
    <div
      class="za-swipeaction-button theme-error"
    >
      <div
        class="za-swipeaction-text"
      >
        右按钮1
      </div>
    </div>
    <div
      class="za-swipeaction-button theme-success"
    >
      <div
        class="za-swipeaction-text"
      >
        右按钮2
      </div>
    </div>
  </div>
  <div
    class="za-swipeaction-content"
    style="-webkit-transition-duration:undefinedms;transition-duration:undefinedms;-webkit-transform:translate3d(0px, 0, 0);transform:translate3d(0px, 0, 0);"
  >
    <div>
      左右都能滑动
    </div>
  </div>
</div>
`;
